src/ucp/proto/proto_select.c

```diff
@@ -0,0 +1,336 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "proto.h"
+#include "proto_select.inl"
+
+#include <ucp/core/ucp_context.h>
+#include <ucp/core/ucp_worker.h>
+#include <ucp/dt/dt.h>
+
+
+/* Parameters structure for initializing protocols for a selection parameter */
+typedef struct {
+    const ucp_proto_select_param_t *select_param;
+    ucp_proto_id_mask_t            mask;
+    ucp_proto_caps_t               caps[UCP_PROTO_MAX_COUNT];
+    void                           *priv_buf;
+    size_t                         priv_offsets[UCP_PROTO_MAX_COUNT];
+} ucp_proto_select_init_protocols_t;
+
+
+static ucs_status_t
+ucp_proto_select_init_protocols(ucp_worker_h worker,
+                                ucp_worker_cfg_index_t ep_cfg_index,
+                                ucp_worker_cfg_index_t rkey_cfg_index,
+                                const ucp_proto_select_param_t *select_param,
+                                ucp_proto_select_init_protocols_t *proto_init)
+{
+    ucp_proto_init_params_t init_params;
+    ucs_string_buffer_t strb;
+    size_t priv_size, offset;
+    ucp_proto_id_t proto_id;
+    ucs_status_t status;
+    void *tmp;
+
+    ucs_assert(ep_cfg_index != UCP_WORKER_CFG_INDEX_NULL);
+
+    init_params.worker        = worker;
+    init_params.select_param  = select_param;
+    init_params.ep_config_key = &worker->ep_config[ep_cfg_index].key;
+
+    if (rkey_cfg_index == UCP_WORKER_CFG_INDEX_NULL) {
+        init_params.rkey_config_key = NULL;
+    } else {
+        init_params.rkey_config_key = &worker->rkey_config[rkey_cfg_index].key;
+
+        /* rkey configuration must be for the same ep */
+        ucs_assertv_always(
+                init_params.rkey_config_key->ep_cfg_index == ep_cfg_index,
+                "rkey->ep_cfg_index=%d ep_cfg_index=%d",
+                init_params.rkey_config_key->ep_cfg_index, ep_cfg_index);
+    }
+
+    proto_init->select_param = select_param;
+    proto_init->mask         = 0;
+
+    /* Initialize protocols and get their capabilities */
+    proto_init->priv_buf = ucs_malloc(ucp_protocols_count * UCP_PROTO_PRIV_MAX,
+                                      "ucp_proto_priv");
+    if (proto_init->priv_buf == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
+    offset = 0;
+    for (proto_id = 0; proto_id < ucp_protocols_count; ++proto_id) {
+        init_params.priv      = UCS_PTR_BYTE_OFFSET(proto_init->priv_buf,
```

Comment:
how is this space used by protocol?

Reply:
(discussed offline) the protocol can keep specific data there such as list of lanes, max_frag for each lane, etc.